### PR TITLE
Finally! a streaming CloudTrail parser

### PR DESCRIPTION
**Pull request by alxarch**
## Background

Why are you making this change? Please provide a reference to any related issues and PRs

## Changes

- List your changes here in more detail

## Testing

- How did you test your change?

### CODE REVIEW

**Comment by 3nvi** [comment]
Yeah! Finally!!!

File: internal/log_analysis/log_processor/processor/logstream/jsonarray.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	} else if size < MinBufferSize {
+		size = MinBufferSize
+	}
+	return &JSONArrayStream{
+		iter: jsoniter.Parse(jsoniter.ConfigDefault, r, size),
+		seek: path,
+	}
+}
+
+type JSONArrayStream struct {
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/processor/logstream/jsonarray.go:42]: Can you add some documentation explaining where this is used?

File: internal/log_analysis/log_processor/sources/s3.go [hunk 2]
```diff
@@ -125,37 +129,46 @@ func shouldIgnoreS3Object(s3Object *S3ObjectInfo) bool {
...
 
 func buildStream(ctx context.Context, s3Object *S3ObjectInfo) (*common.DataStream, error) {
-	s3Client, sourceInfo, err := getS3Client(s3Object.S3Bucket, s3Object.S3ObjectKey)
+	key, bucket := s3Object.S3ObjectKey, s3Object.S3Bucket
+	s3Client, src, err := getS3Client(bucket, key)
 	if err != nil {
-		err = errors.Wrapf(err, "failed to get S3 client for s3://%s/%s",
-			s3Object.S3Bucket, s3Object.S3ObjectKey)
+		err = errors.Wrapf(err, "failed to get S3 client for s3://%s/%s", bucket, key)
 		return nil, err
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/sources/s3.go:136]: nit: just `return nil, errors.Wrap(....)`

File: internal/log_analysis/log_processor/processor/logstream/jsonarray.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import (
+	"fmt"
+	"io"
+	"strconv"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/pkg/errors"
+)
+
+func NewJSONArrayStream(r io.Reader, size int, path ...string) *JSONArrayStream {
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/processor/logstream/jsonarray.go:30]: Can you add documentation explaining what the parameters mean?

File: internal/log_analysis/log_processor/processor/logstream/jsonarray.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+				return seekJSONPath(iter, path)
+			}
+			iter.Skip()
+		}
+		if iter.Error == nil {
+			iter.ReportError(opName, fmt.Sprintf("key %q not found", seek))
+		}
+		return false
+	case jsoniter.ArrayValue:
+		n, err := strconv.ParseInt(seek, 10, 64)
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/processor/logstream/jsonarray.go:112]: Can you add some comments in this method in general? Especially explaining when the code will execute each of the cases.

File: internal/log_analysis/log_processor/sources/s3.go [hunk 5]
```diff
@@ -282,3 +295,10 @@ type SnsNotification struct {
 	events.SNSEntity
 	Token *string `json:"Token"`
 }
+
+// nolint:lll
+var rxCloudTrailLog = regexp.MustCompile(`^(?P<account>\d{12})_CloudTrail_(?P<region>[a-z]{2}-[a-z]+-\d+)_(?P<ts>\d{8}T\d{4}Z)_\w+.json.gz$`)
+
+func isCloudTrailLog(key string) bool {
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/sources/s3.go:302]: can you add some tests for this method?

File: internal/log_analysis/log_processor/sources/s3.go [hunk 6]
```diff
@@ -282,3 +295,10 @@ type SnsNotification struct {
 	events.SNSEntity
 	Token *string `json:"Token"`
 }
+
+// nolint:lll
+var rxCloudTrailLog = regexp.MustCompile(`^(?P<account>\d{12})_CloudTrail_(?P<region>[a-z]{2}-[a-z]+-\d+)_(?P<ts>\d{8}T\d{4}Z)_\w+.json.gz$`)
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/sources/s3.go:300]: 1. Local zones are like `us-west-2-lax-1a`. Will this work?
2. Worth adding a link to the docs here: https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-log-file-examples.html

File: internal/log_analysis/log_processor/parsers/awslogs/cloudtrail.go [hunk 7]
```diff
@@ -118,42 +116,41 @@ var _ parsers.Interface = (*cloudTrailParser)(nil)
...
+	// This is just in case our special detections in the log processor don't work
+	hybrid := hybridCloudTrail{}
+	if err := jsoniter.UnmarshalFromString(log, &hybrid); err != nil {
+		return nil, errors.Wrap(err, "failed to read CloudTrail record JSON")
+	}
+	if hybrid.Records == nil {
+		// The log entry was a single event
+		return p.parseEvent(results, &hybrid.CloudTrail)
+	}
+	for i := range hybrid.Records {
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/awslogs/cloudtrail.go:129]: I really don't like that we have to put this logic here, ideally we could switch the CloudTrail parser to the new parser logic and get rid of custom Go code. If you want to be extra cautious, i think it's fine to keep this for a bit, but maybe log an error if this ever happens - so at least we get notified to fix the regex. Personally I would be fine removing it all together.

Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/awslogs/cloudtrail.go:129]: I think that we need the fallback for cases were log files are delivered in arbitrary paths that do not adhere to the 'official' pattern.

Comment (kostaspap) [reply internal/log_analysis/log_processor/parsers/awslogs/cloudtrail.go:129]: Ok, fine to be extra careful for now

File: internal/log_analysis/log_processor/processor/logstream/jsonarray.go (hunk 3 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/processor/logstream/jsonarray.go:34]: Added

File: internal/log_analysis/log_processor/processor/logstream/jsonarray.go (hunk 4 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/processor/logstream/jsonarray.go:135]: Added

File: internal/log_analysis/log_processor/sources/s3.go (hunk 5 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/sources/s3.go:304]: Added

File: internal/log_analysis/log_processor/sources/s3.go (hunk 6 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/sources/s3.go]: Added link and changed pattern for the local regions

File: internal/log_analysis/log_processor/sources/s3.go (hunk 2 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/sources/s3.go:136]: Do you mean to not display the s3 URL in the error?

File: internal/log_analysis/log_processor/processor/logstream/jsonarray.go (hunk 1 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/processor/logstream/jsonarray.go:48]: Added

File: internal/log_analysis/log_processor/processor/logstream/jsonarray.go (hunk 3 above)
Comment (kostaspap) [reply internal/log_analysis/log_processor/processor/logstream/jsonarray.go:34]: Cool. If I understand the path elements needs to be like `"field name", "<some index>", "field name", "<index>", ...` correct? Might worth pointing it out.

**Review by kostaspap — APPROVED** [review]

**Pull request merged**